hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/amrmproxy/FederationInterceptor.java

```diff
@@ -260,6 +261,16 @@ public class FederationInterceptor extends AbstractRequestInterceptor {
 
   private final MonotonicClock clock = new MonotonicClock();
 
+  /*
+   * For UAM, keepContainersAcrossApplicationAttempts is always true.
+   * When re-register to RM, RM will clear node set and regenerate NMToken for transferred
+   * container. But If keepContainersAcrossApplicationAttempts of MA is false, AM may not
+   * called getNMTokensFromPreviousAttempts, so the NMToken which is pass from
+   * RegisterApplicationMasterResponse will be missing. Here we cache these NMToken,
+   * then pass to AM in allocate stage.
+   * */
+  private Set<NMToken> nmTokenMapFromRegisterSecondaryCluster;
```

Comment:
Using Set<NMToken> is feasible, but should we consider using Map<SubClusterId, Set<NMToken>> for better differentiation of NMToken lists for each subcluster?

Reply:
Set is enough for current implement. Of course we can use Map<SubClusterId, Set>, this maybe useful for some other future issue. I will fixed it.